[system] Remove `theme` & `isRtl` from `useThemeProps`

mnajdova opened the pull request:
**BREAKING CHANGE**

The `isRtl` and `theme` props are no longer added by the `useThemeProps` hook. You can use the `useTheme` hook for this.

```diff
-import { unstable_useThemeProps as useThemeProps } from '@material-ui/core/styles';
+import { unstable_useThemeProps as useThemeProps, useTheme } from '@material-ui/core/styles';

 const Component = (inProps) => {
-  const { isRtl, theme, ...props } = useThemeProps({ props: inProps, name: 'MuiComponent' });
+  const props = useThemeProps({ props: inProps, name: 'MuiComponent' });

+  const theme = useTheme();
+  const isRtl = theme.direction === 'rtl';
 //.. rest of the code
}
```

oliviertassinari commented on packages/material-ui/src/Slider/Slider.js:
```diff
@@ -358,6 +359,10 @@ const shouldSpreadStyleProps = (Component) => {
 
 const Slider = React.forwardRef(function Slider(inputProps, ref) {
   const props = useThemeProps({ props: inputProps, name: 'MuiSlider' });
+
+  const theme = useTheme();
```
Did you consider this API?

```suggestion
  const { props, theme } = useThemeProps({ props: inputProps, name: 'MuiSlider' });
```

mnajdova replied:
I think it’s simpler to return just props, the theme is not required most of the time, and it can be access trough it’s own hook.

oliviertassinari commented on packages/material-ui-system/src/useThemeProps/useThemeProps.js:
```diff
@@ -2,13 +2,7 @@ import getThemeProps from './getThemeProps';
-  };
+  const theme = useTheme(defaultTheme);
+  const mergedProps = getThemeProps({ theme, name, props });
+  return { ...mergedProps };
```
No need to spend CPU cycles to clone

```suggestion
  return mergedProps;
```

mnajdova replied:
I think that there was complain that we are mutating props without it, but I will double check

oliviertassinari replied:
> I think that there was complain that we are mutating props without it, but I will double check

This warning could be hiding something else. AFAIk getThemeProps already clone it.

siriwatknp replied:
agree to return `mergedProps` directly.

oliviertassinari reviewed (commented):
`SwipeableDrawer` can be updated to use `useThemeProps` what we do there is inconsistent.
For `useMediaQuery`, I think that we could almost say the same.

oliviertassinari commented:
> Regarding the argos difference

Very interesting. The diff on its own looks great. I would argue that this demo shouldn't be using `@material-ui/styles` in the first place as we tried to isolate the package.
Also, I believe that we have introduced `@material-ui/private-theming` to solve the problem in the other way around: developers using the ThemeProvider from the core but using makeStyles.

All good from my end

mnajdova commented:
> The diff on its own looks great. I would argue that this demo shouldn't be using @material-ui/styles in the first place as we tried to isolate the package.
Also, I believe that we have introduced @material-ui/private-theming to solve the problem in the other way around: developers using the ThemeProvider from the core but using makeStyles.
All good from my end

Will update the example to use the correct `ThemeProvider`

siriwatknp reviewed (approved):
only the comment that should be fixed, the rest looks good.

siriwatknp commented on packages/material-ui/src/TabScrollButton/TabScrollButton.js:
```diff
@@ -53,8 +54,10 @@ const TabScrollButton = React.forwardRef(function TabScrollButton(inProps, ref)
   const props = useThemeProps({ props: inProps, name: 'MuiTabScrollButton' });
   const { className, direction, orientation, disabled, ...other } = props;
 
-  // TODO: convert to simple assignment after the type error in defaultPropsHandler.js:60:6 is fixed
```
We don't need this anymore?

mnajdova replied:
Yeah it was removed :)

mnajdova commented:
> SwipeableDrawer can be updated to use useThemeProps what we do there is inconsistent.
> For useMediaQuery, I think that we could almost say the same.

Updated and removed the export from for `unstable_getThemeProps`.

eps1lon commented:
> @eps1lon Any idea why it wasn't reported in #26701 (comment)?

It was?

https://mui-dashboard.netlify.app/size-comparison?buildId=28829&baseRef=next&baseCommit=9acf8be6d2fd08210665c294830df3c85c014214&prNumber=26701 and https://mui-maintainer-dashboard-remix.vercel.app/size-comparison?baseCommit=9acf8be6d2fd08210665c294830df3c85c014214&buildId=28829&prNumber=26701&baseRef=next

oliviertassinari commented:
@eps1lon Ok, my bad. I should have been more precise in my question. "reported" as reported in the highlight summary of the mui-pr-bot comment. My assumption is that we check these details links if we see something off in the summary or are working on bundle size-related problems.

eps1lon commented:
> Should we have expected?

That would be my question as well. Did we ever report increases of submodules?

mnajdova commented:
Should we revert to using `getThemeProps`? We should again export this util from `@material-ui/system`.

oliviertassinari commented:
> That would be my question as well. Did we ever report increases of submodules?

@eps1lon I would vote in favor of having this diff. Why? Because we have previously talked about how the size of a package doesn't really matter with tree-shaking. Submodules can't be tree-shaken, they seem more relevant.
Would the output become too noisy? No idea. I hope that a threshold would help, so no real downsides?

@mnajdova I can think of two options:

1. We use `getThemeProps` until we have an unstyled version of useMediaQuery, free of the bundle-size overhead of the default values of the theme
2. We ignore the issue, moving to use the default theme of the system gets us from 20kB gzipped to 8kB gzipped. It's still too much to be relevant with libraries like [react-use](https://github.com/streamich/react-use) but maybe we don't have to target it anyway. But we have to update the docs, saying 1kB is no longer true.

I would vote for 1. to maximize adoption.

mnajdova commented:
> I would vote for 1. to maximize adoption.

Let's go with option 1 then. I guess we will also need a custom `useTheme` hook that does not provide any default value as well.
